interchain-exchange/tutorial/00-tutorial.md

```diff
@@ -0,0 +1,2585 @@
+* Upon receiving the acknowledgement on the source chain
+  * Create a new order book for sell orders
+
+## `SendCreatePair` Message Handling
+
+`SendCreatePair` function was created during the IBC packet scaffolding. Currently, it creates an IBC packet, populates it with source and target denoms and transmits this packet over IBC. Add the logic to check for an existing order book for a particular pair of denoms.
+
+```go
+// x/ibcdex/keeper/msg_server_createPair.go
+import "errors"
+
+//...
+
+func (k msgServer) SendCreatePair(goCtx context.Context, msg *types.MsgSendCreatePair) (*types.MsgSendCreatePairResponse, error) {
+  ctx := sdk.UnwrapSDKContext(goCtx)
+  // Get an order book index
+  pairIndex := types.OrderBookIndex(msg.Port, msg.ChannelID, msg.SourceDenom, msg.TargetDenom)
+  // If an order book is found, return an error
+  _, found := k.GetSellOrderBook(ctx, pairIndex)
+  if found {
+    return &types.MsgSendCreatePairResponse{}, errors.New("the pair already exist")
+  }
+  // ...
+}
+```
+
+## Lifecycle of an IBC Packet
+
+During a successful transmission, an IBC packet goes through 4 stages:
+
+1. Message processing before packet transmission (on the source cahin)
+2. Reception of a packet (on the target chain)
+3. Acknowledgment of a packet (on the source chain)
+4. Timeout of a packet (on the source chain)
+
+In the following section you'll be implementing packet reception logic in the `OnRecvCreatePairPacket` function and packet acknowledgement logic in the `OnAcknowledgementCreatePairPacket` function. Timeout function will be left empty.
+
+## `OnRecv`
+
+On the target chain when an IBC packet is recieved, the module should check whether a book already exists, if not, create a new buy order book for specified denoms.
+
+```go
+// x/ibcdex/keeper/createPair.go
+func (k Keeper) OnRecvCreatePairPacket(ctx sdk.Context, packet channeltypes.Packet, data types.CreatePairPacketData) (packetAck types.CreatePairPacketAck, err error) {
+  // ...
+  // Get an order book index
+  pairIndex := types.OrderBookIndex(packet.SourcePort, packet.SourceChannel, data.SourceDenom, data.TargetDenom)
+  // If an order book is found, return an error
+  _, found := k.GetBuyOrderBook(ctx, pairIndex)
+  if found {
+    return packetAck, errors.New("the pair already exist")
+  }
+  // Create a new buy order book for source and target denoms
+  book := types.NewBuyOrderBook(data.SourceDenom, data.TargetDenom)
+  // Assign order book index
+  book.Index = pairIndex
+  // Save the order book to the store
+  k.SetBuyOrderBook(ctx, book)
+  return packetAck, nil
+}
+```
+
+Define `NewBuyOrderBook` creates a new buy order book.
+
+```go
+ // x/ibcdex/types/buy_order_book.go
+func NewBuyOrderBook(AmountDenom string, PriceDenom string) BuyOrderBook {
+	book := NewOrderBook()
+	return BuyOrderBook{
+		AmountDenom: AmountDenom,
+		PriceDenom: PriceDenom,
+		Book: &book,
+	}
+}
+```
+
+Modify the `buyOrderBook.proto` file to add an order book to the buy order book.
```

Comment:
Modify the `buy_order_book.proto` file to have the fields for creating a buy order on the order book.